### PR TITLE
[OnnxModelLoader] Adding Support of ReduceL2 Onnx Operator

### PR DESCRIPTION
**Pull request by psubhasc**
Summary: Many computer vision models when represented in onnx use reduceL2 operator, thus support of this operator is essential.

Test Plan: Ninja test was done. Additional Test cases where added for expand operator functionality check.

### CODE REVIEW

**Comment by chadrosier** [comment]
Ping.

File: lib/Importer/ONNXModelLoader.cpp [hunk 1]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+      if (axisValue < 0)
+        axisValue += in.dims().size();
+      shapeAxes.push_back((unsigned_t)axisValue);
+    }
+  } else {
+    shapeAxes.resize(in.dims().size());
+    std::iota(shapeAxes.begin(), shapeAxes.end(), 0);
+  }
+  std::sort(shapeAxes.begin(), shapeAxes.end());
+  llvm::ArrayRef<unsigned_t> axes(shapeAxes);
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: Why is this necessary?

File: lib/Importer/ONNXModelLoader.cpp [hunk 2]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+  if (axes.size() > 1) {
+    auto it = std::unique(shapeAxes.begin(), shapeAxes.end());
+    if (it != shapeAxes.end())
+      RETURN_ERR("Axes values are not unique",
+                 ErrorValue::ErrorCode::MODEL_LOADER_UNSUPPORTED_SHAPE);
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: Also could move this into the if case above and skip it for iota?

File: lib/Importer/ONNXModelLoader.cpp [hunk 3]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+  }
+  bool keepDims = true;
+  if (dict.count("keepdims")) {
+    int keepdims;
+    ASSIGN_VALUE_OR_RETURN_ERR(keepdims, loadInt(dict.at("keepdims")));
+    keepDims = (bool)keepdims;
+  }
+
+  // Reduceadd works only for single axis as of now
+  for (auto it = shapeAxes.rbegin(); it != shapeAxes.rend(); ++it) {
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: As in other PR RE: LLVM style for `for` loops:

`for (auto it = shapeAxes.rbegin(), e = shapeAxes.rend(); it != e; ++it)`

File: lib/Importer/ONNXModelLoader.cpp [hunk 4]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+                 ErrorValue::ErrorCode::MODEL_LOADER_UNSUPPORTED_SHAPE);
+  }
+  bool keepDims = true;
+  if (dict.count("keepdims")) {
+    int keepdims;
+    ASSIGN_VALUE_OR_RETURN_ERR(keepdims, loadInt(dict.at("keepdims")));
+    keepDims = (bool)keepdims;
+  }
+
+  // Reduceadd works only for single axis as of now
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: nit: period.

File: lib/Importer/ONNXModelLoader.cpp [hunk 5]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+  for (auto it = shapeAxes.rbegin(); it != shapeAxes.rend(); ++it) {
+    in = G_.createBatchedReduceAdd(opName, in,
+                                   (llvm::ArrayRef<unsigned_t>){*it});
```
Comment (jfix71) [inline lib/Importer/ONNXModelLoader.cpp]: This looks a little funny -- can you try `llvm::makeArrayRef(*it)`? Or if you prefer, `llvm::ArrayRef<unsigned_t>(*it)`?

File: tests/unittests/OnnxImporterTest.cpp [hunk 6]
```diff
@@ -172,6 +172,89 @@ Error checkConstFoldedOutput(std::string NetFilename,
   return Error::success();
 }
 
+// Import reduceL2
+static void importReduceL2Test(std::string &netFilename,
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: tiny nit: const

File: tests/unittests/OnnxImporterTest.cpp [hunk 7]
```diff
@@ -172,6 +172,89 @@ Error checkConstFoldedOutput(std::string NetFilename,
   return Error::success();
 }
 
+// Import reduceL2
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: Improve comment?

File: tests/unittests/OnnxImporterTest.cpp [hunk 8]
```diff
@@ -172,6 +172,89 @@ Error checkConstFoldedOutput(std::string NetFilename,
...
+static void importReduceL2Test(std::string &netFilename,
+                               std::vector<float> inputValues,
+                               std::vector<dim_t> inputShape,
+                               std::vector<dim_t> outputShape,
+                               std::vector<float> expectedValues) {
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: const ref or `llvm::ArrayRef`

File: tests/unittests/OnnxImporterTest.cpp [hunk 9]
```diff
@@ -172,6 +172,89 @@ Error checkConstFoldedOutput(std::string NetFilename,
...
+  float delta = 1e-08;
+  ExecutionEngine EE{};
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+  PlaceholderBindings bindings;
+  Placeholder *graphOutputVar;
+
+  // Load the .onnxtxt model
+  // ONNXModelLoader onnxLD(netFilename, {"x"}, {&x.getType()}, *F);
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: rm

File: tests/unittests/OnnxImporterTest.cpp [hunk 10]
```diff
@@ -172,6 +172,89 @@ Error checkConstFoldedOutput(std::string NetFilename,
...
+  Type input_type(ElemKind::FloatTy, inputShape);
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: `inputType`

File: tests/unittests/OnnxImporterTest.cpp [hunk 11]
```diff
@@ -172,6 +172,89 @@ Error checkConstFoldedOutput(std::string NetFilename,
...
+  std::vector<dim_t> outputShape = {1, 1, 1};
+  std::vector<float> expectedValues = {5.477226};
+  std::string netFilename(GLOW_DATA_PATH
+                          "tests/models/onnxModels/ReduceL2NoAxis.onnxtxt");
+  importReduceL2Test(netFilename, inputValues, inputShape, outputShape,
+                     expectedValues);
+}
+
+/// Test loading reduceL2 op from an ONNX model
+/// with negative axis values
```
Comment (jfix71) [inline tests/unittests/OnnxImporterTest.cpp]: nit: period.

File: tests/unittests/OnnxImporterTest.cpp (hunk 7 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: I will remove it since it is unnecessary (following "testImportPRelu" case)

File: lib/Importer/ONNXModelLoader.cpp [hunk 12]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+  ASSIGN_VALUE_OR_RETURN_ERR(in, getNodeValueByName(op.input(0)));
+
+  // square
+  in = G_.createPow(opName, in, 2);
+
+  // reduceAdd
+  std::vector<unsigned_t> shapeAxes = {};
+  if (dict.count("axes")) {
+    for (int32_t axisValue : dict.at("axes")->ints()) {
+      if (axisValue < 0)
```
[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Incorporated.

[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Included the suggested changes.

[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Incorporate this suggestion.

File: lib/Importer/ONNXModelLoader.cpp [hunk 13]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
   return Error::success();
 }
 
+Error ONNXModelLoader::loadReduceL2(const ONNX_NAMESPACE::NodeProto &op,
+                                    const ArgumentDictionaryTy &dict) {
+  const std::string &opName = loadOperatorName(op);
+  NodeValue in;
+  ASSIGN_VALUE_OR_RETURN_ERR(in, getNodeValueByName(op.input(0)));
+
+  // square
```
[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Incorporate this suggestion.

File: lib/Importer/ONNXModelLoader.cpp [hunk 14]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+    int keepdims;
+    ASSIGN_VALUE_OR_RETURN_ERR(keepdims, loadInt(dict.at("keepdims")));
+    keepDims = (bool)keepdims;
+  }
+
+  // Reduceadd works only for single axis as of now
+  for (auto it = shapeAxes.rbegin(); it != shapeAxes.rend(); ++it) {
+    in = G_.createBatchedReduceAdd(opName, in,
+                                   (llvm::ArrayRef<unsigned_t>){*it});
+    if (keepDims)
```
[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Incorporate this suggestion.

File: lib/Importer/ONNXModelLoader.cpp [hunk 15]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+      in = G_.createExpandDims(opName, in, *it);
+  }
+
+  // sqrt
```
[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Have removed this comment

File: lib/Importer/ONNXModelLoader.cpp [hunk 16]
```diff
@@ -1574,6 +1574,57 @@ Error ONNXModelLoader::loadSpaceToDepth(const ONNX_NAMESPACE::NodeProto &op,
...
+    for (int32_t axisValue : dict.at("axes")->ints()) {
+      if (axisValue < 0)
+        axisValue += in.dims().size();
+      shapeAxes.push_back((unsigned_t)axisValue);
+    }
+  } else {
+    shapeAxes.resize(in.dims().size());
+    std::iota(shapeAxes.begin(), shapeAxes.end(), 0);
+  }
+  std::sort(shapeAxes.begin(), shapeAxes.end());
```
[the remark this answers is not in the document]
Comment (psubhasc) [reply lib/Importer/ONNXModelLoader.cpp]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 8 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.

File: tests/unittests/OnnxImporterTest.cpp (hunk 9 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Removed.

File: tests/unittests/OnnxImporterTest.cpp (hunk 10 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Incorporate this suggestion.

File: tests/unittests/OnnxImporterTest.cpp (hunk 11 above)
Comment (psubhasc) [reply tests/unittests/OnnxImporterTest.cpp]: Done.
